temporal-testing/src/main/java/io/temporal/testing/IdempotentTimeLocker.java

```diff
@@ -0,0 +1,70 @@
+/*
+ *  Copyright (C) 2020 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.testing;
+
+import io.grpc.Context;
+import io.temporal.api.testservice.v1.LockTimeSkippingRequest;
+import io.temporal.api.testservice.v1.UnlockTimeSkippingRequest;
+import io.temporal.serviceclient.TestServiceStubs;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Used to ensure that multiple TimeLockingWorkflowStubs that are blocked at the same time from
+ * multiple threads execute unlock only once and the lock only once.
+ */
+class IdempotentTimeLocker {
+  private final TestServiceStubs testServiceStubs;
+  private final AtomicInteger count = new AtomicInteger(1);
+
+  IdempotentTimeLocker(TestServiceStubs testServiceStubs) {
+    this.testServiceStubs = testServiceStubs;
+  }
+
+  public void lockTimeSkipping(String caller) {
+    if (count.incrementAndGet() == 1) {
+      Context.ROOT.run(
+          () -> {
+            // we want to ignore the gRPC deadline already existing in the context when we
+            // communicate with the test server here to
+            // 1. make sure that this operation is actually performed and
+            // 2. more importantly, don't override and hide any underlying exceptions
+            testServiceStubs
+                .blockingStub()
+                .lockTimeSkipping(LockTimeSkippingRequest.newBuilder().setCallerId(caller).build());
+          });
+    }
+  }
+
+  public void unlockTimeSkipping(String caller) {
```

Comment:
If there are other clients not using this idempotent locker, when this returns time also might not be un-locked. Is that a potential issue? Is the expectation that, if you're using this class, every lock/unlock goes through this class?

Reply:
No, it's not a potential issue. Time is not guaranteed to be unlocked when this method returns. 

This IdempotentLocker only guarantees that time will be locked/unlocked once for a WorkflowStub (see TimeLockingInterceptor where it's used), it doesn't guarantee that the whole test process locks/unlocks only once. Quite opposite, time locking feature is counting, so it's explicitly supported that it can be locked several times. And we actually use it. For example, to disable time locking by default, we just call lockTimeSkipping WITHOUT a correspondent unlock (see https://github.com/temporalio/sdk-java/pull/1076/files#diff-ce2ffc1d1e2ae7b11577efed9e362e7148f6304a85da5e78ba6b3b45a896f15eR101)